keps/sig-scheduling/20200612-simplified-topology-manager.md

```diff
@@ -0,0 +1,229 @@
+  - [Accessing NodeResourceTopology CRD](#accessing-noderesourcetopology-crd)
+- [Use cases](#use-cases)
+- [Test plans](#test-plans)
+- [Graduation criteria](#graduation-criteria)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature enablement and rollback](#feature-enablement-and-rollback)
+- [Implementation history](#implementation-history)
+<!-- /toc -->
+
+# Summary
+
+This document describes behavior of the Kubernetes Scheduler which take into
+account worker node NUMA topology.
+
+# Motivation
+
+After Topology Manager was introduced the problem of launching pod in the
+cluster where worker nodes have different NUMA topology and different amount
+of resources in that topology became actual. Pod could be scheduled on the node
+where total amount of resources are enough, but resource distribution could not
+satisfy the appropriate Topology policy. In this case the pod failed to start. Much
+better behaviour for scheduler would be to select appropriate node where kubelet admit
+handlers may pass.
+
+
+## Goals
+
+- Make scheduling process more precise when we have NUMA topology on the
+worker node.
+
+## Non-Goals
+
+- Change the PodSpec to allow requesting a specific node topology manager policy
+- This Proposal requires exposing NUMA topology information. This KEP doesn't
+describe how to expose all necessary information it just declare what kind of
+information is necessary.
+
+# Proposal
+Kube-scheduler builtin plugin will be added to the main tree. This plugin
+implements a simplified version of TopologyManager it’s different from original topology manager algorithm.
+Plugin checks the ability to run pod only in case of single-numa-node policy on the
+node, since it is the most strict policy, it implies that the launch on the node with
+other existing policies will be successful if the condition for single-numa-node policy passed for the worker node.
+Proposed plugin will use [CRD][1] to identify which topology policy is enabled on the node.
+To work, this plugin requires topology information of the available resource on the worker nodes.
+
+## Topology format
+
+Available resources with topology of the node should be stored in CRD. Format of the topology described
+[in this document][1].
+
+The daemon which runs outside of the kubelet will collect all necessary information on running pods, based on allocatable resources of the node and consumed resources by pods it will provide available resources in CRD, where one CRD instance represents one worker node. The name of the CRD instance is the name of the worker node.
+
+## Plugin implementation details
+
+Since topology of the node is stored in the CRD, kube-scheduler should be subscribed for updates of appropriate CRD type. Kube-scheduler will use informers which will be generated with the name NodeTopologyInformer. NodeTopologyInformer will run in TopologyMatch plugin.
```

Comment:
IIRC, for scheduling pods to nodes, the scheduler avoids races between scheduling decisions by maintaining an in-memory cache of pods assigned to nodes, and updates that after each pod is bound.  However, for this design, there is the possibility that the NodeTopologyMap will not yet be updated with the results of the previous pod's topology decision by the kubelet.  If users schedule multiple pods at once requesting aligned resources, which is common, they will still be scheduled to nodes that can't satisfy their topology requirements.

Are we doing anything to prevent these races?  If not, we should mention this as a known issue, and explain what the expected behavior is.

Reply:
What we can do  in current design (architecture) it's track down the situation when only one numa node was available for allocation, in this case it's the same as non topology based resources.
For cases when more than one of numa nodes available for allocation, if all treated as *used* in this case we can get unallocated resources for a while on this node. Or minimize score of this worker node.
In any case, I agree, it should be mentioned.

Reply:
making sure I understand, the impact of this is a cluster that is initially empty may result in poor placement decisions, but as the cluster increases utilization, the rate of poor scheduling outcomes should decline.

Reply:
I think the drawback discussed in this thread mostly depends on pods launch rate, rather than cluster utilization. More pod launched in the same time, more poor scheduling outcomes.